### ISSUE TITLE
different result

### ISSUE THREAD

**Issue by oceanzhf** [ultralytics/ultralytics#4946]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Validating neu-experiments\train23\weights\best.pt...
Ultralytics YOLOv8.0.175  Python-3.9.13 torch-1.13.0+cu116 CUDA:0 (NVIDIA RTX A4500, 20470MiB)
YOLOv8s summary (fused): 168 layers, 11127906 parameters, 0 gradients, 28.4 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:01<00:00
                   all        180        439      0.708      0.765      0.777      0.447
               crazing        180         76      0.495      0.434      0.455      0.195
             inclusion        180        123      0.827      0.856      0.911      0.494
               patches        180         97      0.831      0.912      0.941      0.652
        pitted_surface        180         34      0.765      0.794      0.805      0.496
       rolled-in_scale        180         49      0.488      0.673      0.611      0.285
             scratches        180         60      0.841      0.917      0.937      0.564
Speed: 0.5ms preprocess, 2.4ms inference, 0.0ms loss, 0.5ms postprocess per image
Results saved to neu-experiments\train23
 Learn more at https://docs.ultralytics.com/modes/train

(ewyolov8) E:\anaconda3\Lib\site-packages\ultralytics>yolo cfg=cfg/default.yaml
Overriding E:\anaconda3\Lib\site-packages\ultralytics\cfg\default.yaml with cfg/default.yaml
Ultralytics YOLOv8.0.175  Python-3.9.13 torch-1.13.0+cu116 CUDA:0 (NVIDIA RTX A4500, 20470MiB)
YOLOv8s summary (fused): 168 layers, 11127906 parameters, 0 gradients, 28.4 GFLOPs
val: Scanning E:\anaconda3\Lib\site-packages\ultralytics\NEU-DET\test1\labels.cache... 180 images, 0 backgrounds, 0 corrupt
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00
                   all        180        439      0.772      0.687      0.758      0.416
               crazing        180         76      0.626      0.224      0.427      0.169
             inclusion        180        123      0.831      0.789      0.876      0.461
               patches        180         97      0.874      0.876      0.917      0.602
        pitted_surface        180         34      0.783      0.706      0.765      0.452
       rolled-in_scale        180         49      0.646      0.612      0.632      0.277
             scratches        180         60      0.874      0.917      0.929      0.537
Speed: 0.5ms preprocess, 3.0ms inference, 0.0ms loss, 1.0ms postprocess per image
Results saved to neu-experiments\val
 Learn more at https://docs.ultralytics.com/modes/val

### Additional

The results of the validation at the end of training are not the same as the results of using best.pt alone to validate the same validation set

**Comment by glenn-jocher** [comment]
@oceanzhf firstly, I'd like to acknowledge your thorough and detailed explanation of the discrepancy in validation results. It definitely helps in understanding the problem better.

The expected behavior should indeed be consistent results between validating at the end of training and using best.pt to validate the same set. However, there are a few factors that can cause discrepancies and I'd recommend checking these:

1. **Random Factors**: Ensure there is no randomness involved in the validation that could lead to different results each time you run the validation. This includes shuffling of data and other random processes that may affect the outcome.

2. **Model State**: Re-check that the model weights and all parameters are exactly the same when the validation is being performed in each situation. Even small differences in initialisation or learning state can impact the results.

3. **Validation Set**: Confirm that the validation set being used is identical in both situations. Any variations or changes in the validation dataset could directly affect the validation results.

4. **Data Preprocessing**: Your data preprocessing steps should be consistent across training and validation. If there's any difference in how the data is prepared or processed, it could lead to differing results.

Please review these factors in your setup. Meanwhile, I'll also look more deeply into the operation of validation routines in YOLOv8 to see if there could potentially be a systematic reason this is occurring. Your findings from the review will be beneficial in this investigation.

Further communication on this issue will help us better understand and resolve this discrepancy. Thanks for your contribution in helping to improve the Ultralytics YOLOv8.

**Comment by thomaskleiven** [comment]
Seems like this is the problem as discussed in [#4536](https://github.com/ultralytics/ultralytics/issues/4536). 

@oceanzhf did you find a solution to the problem?

**Comment by glenn-jocher** [comment]
Hello,

Yes, this seems closely related to issue #4536 that deals with discrepancies between validation results at the end of training and standalone validation using the best.pt model.

I recommended @thomaskleiven to re-verify various aspects, including model state, validation set, randomness during validation, and data preprocessing consistency. The idea is to ensure that there's no variation in any of these factors between the two validation scenarios, as this could cause differing outcomes due to the sensitivity of model performance to these factors.

I'm still pending a response on whether any of these checks provided a solution. To keep things consolidated, I would recommend monitoring issue #4536 for any updates or conclusions.

I truly appreciate your patience in this matter. In the meantime, I'll continue evaluating if there's a systematic issue within the YOLOv8 validation routines that might be causing this.

Best regards,
Glenn
